feg/gateway/services/n7_n40_proxy/n7/model_conversion.go

```diff
@@ -0,0 +1,567 @@
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) *protos.RedirectInformation {
+	if pccRule.RefTcData != nil && len(*pccRule.RefTcData) > 0 && policyDecision.TraffContDecs != nil {
+		tcData, found := policyDecision.TraffContDecs.Get((*pccRule.RefTcData)[0])
+		if found {
+			return GetRedirectInfoProto(tcData.RedirectInfo)
+		}
+	}
+	return nil
+}
+
+func getQosDataFromPccRule(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) *protos.FlowQos {
+	if pccRule.RefQosData != nil && len(*pccRule.RefQosData) > 0 && policyDecision.QosDecs != nil {
+		qosData, found := policyDecision.QosDecs.Get((*pccRule.RefQosData)[0])
+		if found {
+			return GetFlowQosProto(&qosData)
+		}
+	}
+	return nil
+}
+
+func getFlowListFromPccRule(pccRule *n7_sbi.PccRule) []*protos.FlowDescription {
+	var flowList []*protos.FlowDescription
+	if pccRule.FlowInfos != nil {
+		for _, flowInfo := range *pccRule.FlowInfos {
+			if flowInfo.FlowDescription != nil {
+				flow, err := policydb.GetFlowDescriptionFromFlowString(string(*flowInfo.FlowDescription))
+				if err != nil {
+					glog.Errorf("Could not get flow for description %s : %s", *flowInfo.FlowDescription, err)
+				} else {
+					flowList = append(flowList, flow)
+				}
+			}
+		}
+	}
+	return flowList
+}
+
+func getTrackingTypeProto(monKeyPresent bool, ratingGrpPresent bool) protos.PolicyRule_TrackingType {
+	if monKeyPresent && ratingGrpPresent {
+		return protos.PolicyRule_OCS_AND_PCRF
+	} else if monKeyPresent && !ratingGrpPresent {
+		return protos.PolicyRule_ONLY_PCRF
+	} else if !monKeyPresent && ratingGrpPresent {
+		return protos.PolicyRule_ONLY_OCS
+	} else {
+		return protos.PolicyRule_NO_TRACKING
+	}
+}
+
+func GetRedirectInfoProto(redirInfo *n7_sbi.RedirectInformation) *protos.RedirectInformation {
+	redirProto := &protos.RedirectInformation{}
+	if redirInfo == nil {
+		return redirProto
+	}
+	if redirInfo.RedirectEnabled != nil {
+		redirProto.Support = protos.RedirectInformation_DISABLED
+		if *redirInfo.RedirectEnabled {
+			redirProto.Support = protos.RedirectInformation_ENABLED
+		}
+	}
+	if redirInfo.RedirectAddressType != nil {
+		redirProto.AddressType = GetRedirectAddrTypeProto(redirInfo.RedirectAddressType)
+	}
+	if redirInfo.RedirectServerAddress != nil {
+		redirProto.ServerAddress = *redirInfo.RedirectServerAddress
+	}
+
+	return redirProto
+}
+
+func GetRedirectAddrTypeProto(
+	redirAddrType *n7_sbi.RedirectAddressType,
+) protos.RedirectInformation_AddressType {
+	if redirAddrType == nil {
+		return protos.RedirectInformation_IPv4
+	}
+	switch *redirAddrType {
+	case n7_sbi.RedirectAddressTypeIPV4ADDR:
+		return protos.RedirectInformation_IPv4
+	case n7_sbi.RedirectAddressTypeIPV6ADDR:
+		return protos.RedirectInformation_IPv6
+	case n7_sbi.RedirectAddressTypeSIPURI:
+		return protos.RedirectInformation_SIP_URI
+	case n7_sbi.RedirectAddressTypeURL:
+		return protos.RedirectInformation_URL
+	default:
+		return protos.RedirectInformation_IPv4
```

Comment:
if this is not supposed to happen you need to trigger an error. At least log it and leave it to 0

Reply:
RedirectInformation_IPv4 is the default value (0). Again here it is getting converted to protobuf and the type is not a pointer. Hence setting to default value when not set in the sbi type.

Reply:
on protos, values are converted into default: 0 if int "" if string. 

But arrays are pointers and they are converted into nil when sending the proto

Reply:
That is correct, but here the proto type is a enum and the go type is an int and hence returning a default value.

Reply:
I think this is hiding a possible error. Can you at least trigger an log message saying that the field was defaulted?